source/ConversationPanel.cpp

```diff
@@ -343,21 +343,27 @@ void ConversationPanel::Goto(int index, int choice)
 void ConversationPanel::Exit()
 {
 	GetUI()->Pop(this);
-	// If this is a conversation offered from boarding or assisting an NPC,
-	// being forced to leave via LAUNCH, FLEE, or DEPART destroys it. If it
-	// is a hostile NPC and not destroyed, and its mission (e.g. "leave me
-	// alone, here's X as payment") was not accepted, open a BoardingPanel
-	// to allow plundering it - unless you're dead.
-	if(node != Conversation::DIE && player.BoardingShip())
+	// Some conversations may be offered from an NPC, e.g. an assisting or
+	// boarding mission's `on offer`, or from completing a mission's NPC
+	// block (e.g. scanning or boarding or killing all required targets).
+	if(node == Conversation::DIE || node == Conversation::EXPLODE)
+		player.Die(node, ship);
+	else if(ship)
 	{
+		// A forced-launch ending (LAUNCH, FLEE, or DEPART) destroys any NPC.
 		if(Conversation::RequiresLaunch(node))
-			player.BoardingShip()->Destroy();
-		else if(player.BoardingShip()->GetGovernment()->IsEnemy()
+			// TODO: This may need a ShipEvent, but not all instances
+			// are necessarily caused by the player.
+			ship->Destroy();
+		// Only show the BoardingPanel for a hostile NPC that is being boarded.
+		// (NPC completion conversations can result from non-boarding events.)
+		else if(ship->Position().Distance(player.Flagship()->Position()) <= 1.
```

Comment:
Could this check `&ship == player.BoardingShip().get()` instead? Or is there a reason that wouldn't work? Relying on the distance to identify a ship you're boarding just seems like a bit of a hack.

Reply:
Oops, never mind, now I remember that the point was to eliminate PlayerInfo::BoardingShip().

Reply:
I wonder if checking `ship->IsYours()` would work. The only times a ConversationPanel has a ship is if 1. it's a flight check message (for one of your ships) or 2. it's a conversation being shown because you boarded a ship. So, if the ship is not yours, I think it must be a ship you're boarding.

Reply:
NPC completion also packages a ship into ConversationPanel (hence the code comments), and the ship isn't necessarily being boarded - it might have been scanned, destroyed, disabled, etc.